utils/device_pool.py

```diff
@@ -1,107 +1,211 @@
+
+class DevicePool(object):
+
+    serial_list = None
+    device_obj_map = None
+    current_device_object = None
+
+    def __init__(self, specify_serial=None):
+        self.serial_list = self._gen_serial_list()
+        self.device_obj_map = self._gen_device_obj_map()
+        if specify_serial and self.chk_serial_exist(specify_serial):
+            self.current_device_object.serial = self.device_obj_map[specify_serial]
+
+    def _gen_device_obj_map(self):
+        obj_map = {}
+        for serial in self.serial_list:
+            tmp_object = DeviceObject(serial)
+            if serial in obj_map.keys():
+                logging.error("Duplicate serial in serial list!!")
+            else:
+                obj_map[serial] = tmp_object
+        return obj_map
+
+    def _gen_serial_list(self):
+        # adb devices here
+        tmp_list = exec_process(['/usr/bin/adb', 'devices']).splitlines()
+        tmp_list.pop(0)  # remove the description from adb
+        serial_list = map(lambda x: x.split("\t")[0], filter(lambda x: x, tmp_list))
+        return serial_list
+
+    def get_device(self):
+        if self.current_device_object:
+            logger.info("Use current device object instead of take new obj!")
+            return self.current_device_object
+        for serial_key in self.device_obj_map.keys():
             try:
-                if not self.lock_list[next_index].i_am_locking():
-                    self.lock_list[next_index].acquire(timeout=60)
-                    self.my_lock = self.lock_list[next_index]
-                    self.current_lock_index = next_index
-                    return self.lock_list[next_index].path
+                if not self.device_obj_map[serial_key].is_file_locked() and not self.device_obj_map[serial_key].is_in_forwarded_list():
+                    self.device_obj_map[serial_key].acquire_file_lock()
+                    self.current_device_object = self.device_obj_map[serial_key]
+                    logger.info("Get device with serial [" + serial_key + "]!")
+                    return self.device_obj_map[serial_key]
             except:
+                # fail to get lock e
+                # ventually, TODO: raise corresponding exception
+                logger.error(traceback.format_exc())
                 logger.error("Failed to get lock!!")
                 pass
+        logger.warning("No available device.  Please retry after device released")
         return None
 
     def release(self):
-        if not self.my_lock:
-            logger.info("No lock acquired")
+        if not self.current_device_object:
+            logger.info("No device in use!")
             return
         try:
-            self.my_lock.release()
+            self.current_device_object.release_file_lock()
         except NotLocked:
-            logger.info("Lock is released")
-        self.my_lock = None
+            logger.info("File lock is released")
+
+        if self.current_device_object.is_in_forwarded_list:
```

Comment:
If Shako changes this function to property, it should be fine. However, line 192 and 195 and other lines needed to be handled.